### PR TITLE
Issue 133 set gridding outputs confusing

### PR DESCRIPTION
**Pull request by skymandr**
We currently know that this has issues, but I am creating a pull request nonetheless to get the proper review process started. Will look at the reported problem and see if I can fix it, but meanwhile would appreciate if someone would look at the changes not directly related to the grid API.

### CODE REVIEW

**Comment by skymandr** [comment]
I would also very much appreciate input on how to write tests for the issue.

**Comment by Endimion75** [comment]
I am currently cleaning up the CCC prototype so it can be used to test the gridding. hopefully I will be able to push the code soon

**Comment by Endimion75** [comment]
I tested the gridding and even though it did not pop an error, the grid seems to be rotated:

sc3

is like the row are now cols and the cols are now rows

**Comment by Endimion75** [comment]
Also Issue 134 remains, but now extended to the set grid API path:

/api/data/ccc/<ccc_identifier>/image/<image_identifier>/plate/<plate_number>/grid/set

Here I also get a bad request page status 400 or in other words an error response from the Ajax call to the API rather than a successful response, with a json object indicating that the gridding was unsuccessful.

**Comment by Endimion75** [comment]
I though that the issue 134 was localized to the compression API call, but now that I saw it in the gridding call I realized that something much have changed in the API code that affected other API paths.  the CCC UI was wired to expect a behaviour that is no more.

It is troubling as belabour as fundamental as this is changing ...

**Comment by Endimion75** [comment]
Regarding issue 134.

When I use the old debug_CORS branch (which is the latest stable branch I am using to continue my development) the problem with the success/error responses on the "set grid" API path doesn't exist.

So it was introduced while attempting to fix the issues in the new pull request and is not as old as I though it to be!  Which is great news!  I don't have to double check all the UI again :D

**Comment by skymandr** [comment]
Regarding 134, 40X with failure description in the JSON response is the new way the API will work. We will successively phase out 200 for failed responses (most API endpoints should be converted already, but some stragglers may remain).

Regarding the core issue, we'll look into it now! =)

**Comment by skymandr** [comment]
We identified the issue with the rotated grid and think we fixed it at its source in the latest commit.

**Comment by local-minimum** [comment]
I'd like @Endimion75 to verify that this update now works better. We expect the grid origin / `[0, 0]` position to be in the lower left corner from now.

**Comment by Endimion75** [comment]
This is what I got:

sc6

The gird is standing on the short side same as the slice, but the grid seems is still off by one row

**Comment by local-minimum** [comment]
That the grid ends up in the wrong position and especially in this direction is weird, but I think that should be a separate issue to not make a too large issue out of this. I'll check the code tomorrow. And btw @Endimion75 if you could make more frequent pull requests for the UI we could check these things out ourselves unless you think this is good workflow that we ask you to test stuff, in that case everything is good.

File: scanomatic/image_analysis/grid_array.py [hunk 1]
```diff
@@ -218,13 +237,13 @@ def grid_cell_size(self):
     @property
     def grid(self):
 
-        return self._grid.tolist()
+        return self._grid[:, ::-1, :].tolist()
```
Comment (local-minimum) [inline scanomatic/image_analysis/grid_array.py]: This will sound weird but perhaps a docstring to explain the reason for inversing the y-axis

**Comment by skymandr** [comment]
Added a docstring. Will work on implementing tests during the week. We need to discuss how to proceed with regard to Issue #134: What problem was `json_abort` supposed to solve, and is it worth it considering the extra work for Luciano?

**Comment by skymandr** [comment]
I would also request that someone with more insight opens an issue for the problem with the inconsistent gridding results.

**Comment by skymandr** [comment]
Regarding #134 and the `json_abort`s - should I revert those changes?

**Comment by local-minimum** [comment]
There are already various api-calls that sends error codes in master since previous commits, and to me this will make the api and UI development much cleaner when using `promise` type  interfaces that are the coming/current standards of EC6 as well as makes the api more restful (if i get how that is supposed to work). As it is now, without sending error-codes basically all calls to the api must check for `success=False` in successful calls and check in failed calls for how to handle that the api did not agree with the call. This should mean code duplication in all api-calls. If we from the API guarantee that if the call did not go through that it will be sent back with an error code, the UI can assume that things have worked well when 200 is returned and must only handle problems in error/fail callbacks. So my vote is clearly for using abort-codes. This by all means must decrease the complexity and needs of the UI.

**Comment by Endimion75** [comment]
@local-minimum:  You do what you need to do.  When done, just document somewhere the new pattern that the API will follow and I will update my end of the code.  All I ask is consistency with the new pattern across the API calls that are used by the CCC UI.  When I come back from holidays in 5 or so weeks, I will implement the new changes in the CCC UI and let you know who it went.

**Comment by skymandr** [comment]
Updated with tests. The test applies a correction of `(3, 0)`, without which the gridding (when it works) looks like this:
gridding_right_correction_none
but with correction, the gridding (when it works) is correct:
gridding_right_correction_ 3 0

I write "when it works" above, since sometimes the needed offset is `(4, 0)`, but I will go into that in more detail in #138.

Conclusion is that `GridArray` works as expected, but that the gridding is sometimes wrong.

File: scanomatic/image_analysis/test/test_grid_array.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            grid_cell = grid_array[(row, col)]
+            try:
+                assert grid_cell.xy1[0] < grid_array.grid[0][row][col]
+                assert grid_cell.xy1[1] < grid_array.grid[1][row][col]
+                assert grid_cell.xy2[0] > grid_array.grid[0][row][col]
+                assert grid_cell.xy2[1] > grid_array.grid[1][row][col]
+            except Exception as err:
+                raise Exception("{} for cell ({}, {})".format(err, row, col))
+
+
```
Comment (local-minimum) [inline scanomatic/image_analysis/test/test_grid_array.py]: Too many lines here

File: scanomatic/image_analysis/test/test_grid_array.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    image = ndimage.io.imread(
+        './scanomatic/image_analysis/test/testdata/test_fixture.tiff')
+    grid_array_instance = grid_array_module.GridArray(
+        image_identifier, pinning, analysis_model)
+    correction = (3, 0)
+    grid_array_instance.detect_grid(image, grid_correction=correction)
+    return grid_array_instance
+
+
+class TestGridDetection():
```
Comment (local-minimum) [inline scanomatic/image_analysis/test/test_grid_array.py]: I would like a test of a grid_cell's identifier and testing that some `grid_cell`s position https://github.com/Scan-o-Matic/scanomatic/blob/master/scanomatic/image_analysis/grid_cell.py#L37 matches the position you request from the `grid_array` like with line 31 below.

**Pull request merged**